gremlin-javascript/src/main/javascript/gremlin-javascript/lib/driver/driver-remote-connection.js

```diff
@@ -151,7 +156,16 @@ class DriverRemoteConnection extends RemoteConnection {
       return;
     }
 
-    if (response.status.code >= 400) {
+    if (response.status.code === responseStatusCode.authenticationChallenge && this._authenticator) {
+      this._authenticator.evaluateChallenge(response).then(res => {
```

Comment:
To avoid leaking those response handlers, you should call `_clearHandler()`:

```javascript
// Similar to how its being done below, maybe we can move that to a separate method
this._clearHandler(response.requestId);
handler.callback(null, { traversers: [] });
```

So the final solution should be something like:
```javascript
this._authenticator.evaluateChallenge(response)
  .then(() => this.submit(null, 'authentication', res))
  .then(() => this._clearHandlerAndInvokeCallback(response.requestId, handler, []))
  .catch(handler.callback);
return;
```

Reply:
On this problem, the issue is that submit creates a new requestId for each message sent. So when sending the auth response a new requestId is created for that authentication response. The Gremlin server responds with the requestId of the previous call so the authentication requestId never gets dealt with. Am I right in thinking that your proposal is to remove all pending handlers from the handler array? Is it safe to assume that there will only ever be one handler at a time in that array, apart from when authentication is happening? Am I making sense?

Reply:
You are right, I was thinking each request have to have their own unique request id, but its not the case for auth.

The requestId that have to be sent should be the same as the one in the response.

Reply:
We should have a way to build a request (in this case an auth request) and send it providing an existing `requestId`, that doesn't add a key to the `this._responseHandlers`. That way the whole flow uses the same `requestId` and the response handler w/ the original callback is maintained until the response message (out of the auth flow) for the original request is obtained.